### PR TITLE
JS Styleguide

### PR DESCRIPTION
**Pull request by olleolleolle**
We are interested in the JS styleguide. Re-opening PR.

This reverts commit bf882c567c16c6844e2b121fcf4b183928f7b5c3, reintroducing the document @aalin began. Thanks! 🌟 

If there are section where we want other language, please comment.

Ciao!

bild

### CODE REVIEW

**Review by tsrandrei — APPROVED** [review]

File: styleguide/javascript/README.md
Comment (henrik) [inline styleguide/javascript/README.md:5]: I would like us to make more specific plans about how to go about this in a productive way – e.g. Calle and I have kept using jQuery for sprinkles of theme JS because that's the path of least resistance in the project currently.

Also, considering the discussion we had about "the next person to do uploads MUST adapt CarrierWave to support it", it might be good to set expectations more explicitly. Current phrasing both seems laid back about it ("avoid") and very forceful ("Just don't use it") :) I would suggest going in the more laid-back direction – encourage The New Way but don't "force" devs to switch focus to rewrite existing code they happen to touch.

Perhaps before we merge this PR, someone could rewrite one or a few of our current sprinkles of jQuery in The Desired Way? It would let us all look at it and (I suspect) agree we like it better, and it would also be something we could even point to in the styleguide as an example of how to do things in future.

Perhaps I could do this with the theme JS we added if I get some spare time today – but no promises :)

File: styleguide/javascript/README.md
Comment (henrik) [inline styleguide/javascript/README.md:31]: So this is just something you add to package.json in the project? Any editor tooling needed as well?

File: styleguide/javascript/README.md
Comment (henrik) [inline styleguide/javascript/README.md:45]: Does Rails still default to yarn? If so, is using npm instead of yarn with Rails straightforward or does it come with "fighting against the conventions" pains?

File: styleguide/javascript/README.md
Comment (henrik) [inline styleguide/javascript/README.md:86]: Which other guide? The code we removed above?

**Review by henrik — CHANGES_REQUESTED** [review]
I'd like us to clarify how to think about the "not jQuery" rule and maybe have at least one example to follow before we make it "law".

File: styleguide/javascript/README.md
Comment (p-wall) [reply styleguide/javascript/README.md:5]: jQuery is adopted still, see https://trello.com/b/ytd5jcWY/tech-radar

Comment (soma) [reply styleguide/javascript/README.md:5]: I'd be happy to update the tech radar if the team is on board with this change.

File: styleguide/javascript/README.md
Comment (olleolleolle) [reply styleguide/javascript/README.md:45]: Short answer: Rails defaults to yarn.

My take: For non-Rails apps, we could be more `npm`?

**Review by soma — APPROVED** [review]

File: styleguide/javascript/README.md
Comment (aalin) [reply styleguide/javascript/README.md:5]: Don't use jQuery in new projects. If jQuery is present in whatever project you're working in, then use it.

Comment (aalin) [reply styleguide/javascript/README.md:5]: In new projects we shouldn't do much DOM-manipulation at all...

File: styleguide/javascript/README.md
Comment (aalin) [reply styleguide/javascript/README.md:31]: Yes, it says in the README how you use the config. Then you need to [configure prettier to run](https://prettier.io/docs/en/precommit.html).

There are probably plugins for all editors, but none are needed. Easiest is to just set up a pre-commit-hook

File: styleguide/javascript/README.md
Comment (aalin) [reply styleguide/javascript/README.md:45]: I'd say, don't change anything, like, if Rails uses yarn, then we use yarn in that project.

If we start a new project, use npm, if we need some feature that exists in yarn but not in npm, use yarn.

To me it just makes more sense to default to using `npm` instead of `yarn` since you have to use npm to install yarn anyways, and they solve the same problem.

File: styleguide/javascript/README.md
Comment (aalin) [reply styleguide/javascript/README.md:86]: https://github.com/barsoom/devbook/blob/master/styleguide/react/README.md

File: styleguide/javascript/README.md
Comment (aalin) [reply styleguide/javascript/README.md:5]: Just don't add jQuery to any new projects. You don't need it.

**Review by aalin — APPROVED** [review]

**Review by p-wall — APPROVED** [review]

**Review by tskogberg — APPROVED** [review]

**Review by joakimk — APPROVED** [review]

File: styleguide/javascript/README.md
Comment (olleolleolle) [reply styleguide/javascript/README.md:35]: In future PRs to the devbook, we can assist users with specific guidance.

File: styleguide/javascript/README.md
Comment (henrik) [reply styleguide/javascript/README.md:5]: Sounds like a good clarification to make. Don't use it in new projects. Fine to keep using in old ones. Might also be nice to encourage writing new code in existing projects this way, just as long as we're clear that there's no requirement to rewrite huge chunks of existing code just because you need to modify it.

I recently rewrote a side project jQuery to vanilla JS and it was mostly fine. Some things got a little worse and there were some initial gotchas (`document.querySelector(".foo").stuff()` is an error if there's no .foo, unlike jQuery) but definitely felt worth it.

File: styleguide/javascript/README.md
Comment (henrik) [reply styleguide/javascript/README.md:49]: Right. Could we make that explicit, then?

File: styleguide/javascript/README.md
Comment (henrik) [reply styleguide/javascript/README.md:90]: I think we should merge them as part of this PR. CoffeeScript is just incidental to that document – it's mostly about React style. Would be nice to have all React styleguides in one place. We could use JS in the examples.

File: styleguide/javascript/README.md
Comment (aalin) [reply styleguide/javascript/README.md]: Removed the entire section.

File: styleguide/javascript/README.md
Comment (aalin) [reply styleguide/javascript/README.md:78]: No. The other guide has `React.createClass` which has been removed from React. Some things would be fixed by prettier/eslint. Other things are already documented in the official React docs.

File: styleguide/javascript/README.md
Comment (olleolleolle) [reply styleguide/javascript/README.md:35]: @aalin I will take the radical step and make those two repositories public, for ease of use.

Comment (olleolleolle) [reply styleguide/javascript/README.md:35]: DONE.

**Comment by henrik** [comment]
I'll merge this and make any adjustments after :)

**Pull request merged**
